File: payroll/frontend/src/api/departments.ts
export interface Department {
  id: number
  code: string
  name: string
  is_active: boolean
}

interface DepartmentsResponse {
  success: boolean
  count: number
  data: Department[]
}

const API_URL =
  import.meta.env.VITE_API_URL || 'http://127.0.0.1:8000/api'

export async function getDepartments(): Promise<Department[]> {
  const response = await fetch(`${API_URL}/departments`)

  if (!response.ok) {
    throw new Error(`โหลดข้อมูลหน่วยงานไม่สำเร็จ: ${response.status}`)
  }

  const result = (await response.json()) as DepartmentsResponse

  if (!result.success) {
    throw new Error('API ไม่สามารถส่งข้อมูลหน่วยงานได้')
  }

  return result.data
}

File: payroll/backend/departments.py
from DBHelper import DBHelper


class Departments:

    def __init__(self):
        self.db = DBHelper()

    def dump(self):
        data, columns = self.db.fetch(
            """
            SELECT
                id,
                code,
                name,
                is_active
            FROM public.departments
            WHERE is_active = TRUE
            ORDER BY id
            """
        )

        departments = []

        for row in data:
            departments.append(dict(zip(columns, row)))

        return departments

    def read(self, department_id):
        data, columns = self.db.fetch(
            """
            SELECT
                id,
                code,
                name,
                is_active
            FROM public.departments
            WHERE id = %s
            """,
            (department_id,)
        )

        if len(data) == 0:
            return (
                {
                    "Is Error": True,
                    "Error Message": (
                        f"ไม่พบหน่วยงานรหัส {department_id}"
                    )
                },
                {}
            )

        department = dict(zip(columns, data[0]))

        return (
            {
                "Is Error": False,
                "Error Message": ""
            },
            department
        )
